spyder/api/plugins.py

```diff
@@ -186,100 +106,164 @@ class SpyderPlugin(BasePlugin):
-        message = ''
-        valid = True
-        return valid, message
+    # --- Private methods ----------------------------------------------------
+    # ------------------------------------------------------------------------
+    def _initialize(self, parent, configuration):
+        self.main = parent
+        self._config = configuration or CONF
 
+        # Filesystem path to the root directory that contains the plugin.
+        self.PLUGIN_PATH = os.path.dirname(inspect.getfile(self.__class__))
 
-# =============================================================================
-# SpyderPluginWidget
-# =============================================================================
-class BasePluginWidget(QWidget, BasePluginWidgetMixin):
-    """
-    Basic functionality for Spyder plugin widgets.
+        # Register plugin configuration
+        if self._config is not None:
+            self._config.register_plugin(self)
 
-    WARNING: Don't override any methods or attributes present here!
-    """
+        # Check plugin compatibility
+        check_compatibility, message = self.check_compatibility()
+        self._is_compatible = True
+        if not check_compatibility:
+            self._is_compatible = False
+            self._show_compatibility_message(message)
 
-    # Signal used to update the plugin title when it's undocked
-    sig_update_plugin_title = Signal()
+        # Connect signals to slots.
+        self.sig_show_status_message.connect(self.show_status_message)
+        self.sig_option_changed.connect(self.set_option)
 
-    def __init__(self, main=None):
-        super(BasePluginWidget, self).__init__(main)
+    def _show_compatibility_message(self, message):
+        """Show a compatibility message."""
+        messageBox = QMessageBox(self)
+        messageBox.setWindowModality(Qt.NonModal)
+        messageBox.setAttribute(Qt.WA_DeleteOnClose)
+        messageBox.setWindowTitle(_('Compatibility Check'))
+        messageBox.setText(message)
+        messageBox.setStandardButtons(QMessageBox.Ok)
+        messageBox.show()
+
+    def _create_configwidget(self, dlg, main_window):
+        """Create configuration dialog box page widget."""
+        if self.CONFIGWIDGET_CLASS is not None:
+            parent = self
+            main = dlg
+            if not hasattr(self, 'dockwidget'):
+                # Prevent QWidget assignment to a plugin that does not have
+                # a graphical widget.
+                parent = dlg
+                main = main_window
+            configwidget = self.CONFIGWIDGET_CLASS(parent, main)
+            configwidget.initialize()
+            return configwidget
+
+    # --- API: available methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+
+    # FIXME: Also we need to decide if we want to use the word plugin on every
+    # method as a way of consistency. This should be the `get_plugin_option`
```

Comment:
I'm -1 on this. If these methods belong to a class called Plugin, then it's clear they refer to plugins.

Reply:
It is not necessary I agree, but some methods have it and some methods do not. I am striving for  consistency which at the end makes stuff predictable for users and developers.

We have `register_plugin`  `get_plugin_name` but then we have methods that do not...

If we add fuzzy matching to spyder (soon to come) doing `SomePlugin.plugin`would provide ALL the methods that the user may use for things so there is a small added benefit here.

We should remove plugin from all things or add it to all but the current state is a messy mixture of different ideas.

Reply:
I see. I'd prefer we remove `plugin` from `register_plugin`, `get_plugin_name`.

Reply:
Removed plugin word